### ISSUE TITLE
Make most of requirements optional and load them lazily

### ISSUE THREAD

**Issue by KOLANICH** [ClimbsRocks/auto_ml#368]
Now auto_ml bloats users setup with the packages which may be unneeded for him.
For example, lightgbm is unneded to the ones using xgboost, h5py is unneded to the ones not using hdf5, dill is unneded to the ones not saving the models. And tensorflow is especially un eeded. sklearn-deap2 also may be unneded.

Also there are some packages like pathos. Do I need them if I gonna train only on my PC?

so, it may worth to:

1 make the modules be loaded lazily.

You can do that, for example, with https://github.com/KOLANICH/backblaze_analytics/blob/master/backblaze_analytics/utils/lazyImport.py

This can also save some time and resources to the people not using all the models available.

2 move them into optional dependencies;

3 change the code in the way to make them not required in the use cases they are not required.

**Comment by ClimbsRocks** [comment]
Thanks for the note! 

I'll address the specific packages first, then the broader question. 

1. lightgbm- we use this for quantile regression
2. h5py- we use this when saving some models
3. dill- yep, not needed if you're not saving models. but saving models is a major use case of auto_ml. if you're only looking for batch processing accuracy, i'd probably recommend [auto-sklearn](https://github.com/automl/auto-sklearn) or [xcessiv](https://github.com/EpistasisLab/tpot) or [tpot](https://github.com/EpistasisLab/tpot). They're better about doing things like ensembling that work well for batch predictions, but add a layer of complexity that's not as obvious of a win in production environments, and make the analytics from models much more difficult to interpret. 
4. tensorflow - we removed this from the list of requirements some releases ago, along with keras. still usable, but now the user has to install it. 
5. sklearn-deap2 - used for our grid search replacement, evolutionary-algorithm search

In general, I think it's important for someone to be able to `pip install auto_ml` and have everything they need to run all the major use cases already installed. Hard drive space is cheap, and outside of tensorflow, most of these packages don't use up much space. But the developer time cost of debugging why something isn't working is expensive. 

One of the goals of this project is to have machine learning "just work". It seems contrary to that goal to not install the packages people need for some pretty major use cases. The packages we rely on are also, unfortunately, not always the best engineering projects. It's been a major time sink for me updating auto_ml as the external packages cut new releases that break their APIs without warning. So specifying which versions of packages we support and installing those versions automatically saves everyone a big headache. 

I see the frustration, and I love getting suggestions, but I think the current approach best suits the aims of this project. 

1. Hard disk space is cheap
2. All of the packages you mentioned (with the exception of tensorflow, which is a beast) are pretty light on memory
3. Developer time is expensive, and I don't want people wondering why we didn't install necessary packages automatically
4. Making ML as turnkey as possible is one of auto_ml's big goals. 

Let me know if there's something major I didn't consider. Again, thanks for the feedback- would love to hear what other ideas you have.

**Comment by KOLANICH** [comment]
>1 Hard drive space is cheap

>hard drive

Not everyone and everything has a hard drive.

2 If a one has several MiB free space he will count every kiB. Offloading the files into clouds is not an option.
